generator/src/main/java/com/graphicsfuzz/generator/mutateapi/FactManager.java

```diff
@@ -0,0 +1,116 @@
+package com.graphicsfuzz.generator.mutateapi;
+
+import com.graphicsfuzz.common.ast.TranslationUnit;
+import com.graphicsfuzz.common.ast.decl.Declaration;
+import com.graphicsfuzz.common.ast.decl.FunctionDefinition;
+import com.graphicsfuzz.common.ast.decl.FunctionPrototype;
+import com.graphicsfuzz.common.ast.decl.Initializer;
+import com.graphicsfuzz.common.ast.decl.VariableDeclInfo;
+import com.graphicsfuzz.common.ast.decl.VariablesDeclaration;
+import com.graphicsfuzz.common.ast.expr.Expr;
+import com.graphicsfuzz.common.ast.expr.FloatConstantExpr;
+import com.graphicsfuzz.common.ast.expr.FunctionCallExpr;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
+import com.graphicsfuzz.common.ast.stmt.BlockStmt;
+import com.graphicsfuzz.common.ast.stmt.DeclarationStmt;
+import com.graphicsfuzz.common.ast.stmt.ExprStmt;
+import com.graphicsfuzz.common.ast.stmt.ReturnStmt;
+import com.graphicsfuzz.common.ast.stmt.Stmt;
+import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.ast.type.QualifiedType;
+import com.graphicsfuzz.common.ast.visitors.StandardVisitor;
+import com.graphicsfuzz.common.util.IRandom;
+import com.graphicsfuzz.common.util.PipelineInfo;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FactManager {
+
+  private final FactManager prototype;
+
+  private final Map<VariableDeclInfo, VariableFact> variableFacts;
+  private final Map<FunctionPrototype, FunctionFact> functionFacts;
+
+  public FactManager(FactManager prototype) {
+    this.prototype = prototype;
+    variableFacts = new HashMap<>();
+    functionFacts = new HashMap<>();
+  }
+
+  public static Expr generateExpr(TranslationUnit tu,
+                                  PipelineInfo pipelineInfo,
+                                  FactManager factManager,
+                                  FunctionDefinition functionDefinition,
+                                  Stmt stmt,
+                                  Value value,
+                                  IRandom generator) {
+
+    String primitiveValueString = value.getData().get(0).get().toString();
+
+    switch (generator.nextInt(2)) {
```

Comment:
In addition to the cases that you have, consider adding the following cases:

- a case that returns a ConstantExpr with the required value - a FloatConstantExpr for a float value, an IntConstantExpr for an int, etc.

- a case that returns a VariableIdentifierExpr using a variable the fact manager knows about (so don't create a new fact, but use an existing fact)

- a case that returns a FunctionCallExpr using a function the fact manager already knows about